merkle-tree/concurrent/src/concurrent_merkle_tree.rs

```diff
@@ -0,0 +1,231 @@
+                let padding = 32 - MAX_HEIGHT;
+                let common_path_len =
+                    ((i ^ changelog_entry.index as usize) << padding).leading_zeros() as usize;
+                let critbit_index = (MAX_HEIGHT - 1) - common_path_len;
+
+                updated_proof[critbit_index] = changelog_entry.path[critbit_index];
+            } else {
+                println!("not swapping critbit");
+                updated_leaf = changelog_entry.path[0];
+            }
+        }
+        println!("about to validate proof");
+        println!("old_leaf: {old_leaf:?}, updated_leaf: {updated_leaf:?}");
+        if is_valid_proof::<H, MAX_HEIGHT>(
+            self.roots[self.current_index as usize],
+            old_leaf,
+            i,
+            proof,
+        )? && updated_leaf == old_leaf
+        {
+            println!("proof valid");
+            self.update_path_to_leaf(new_leaf, i, &updated_proof)?
+        }
+
+        Ok(())
+    }
+
+    pub fn replace_leaf(
+        &mut self,
+        root: [u8; 32],
+        old_leaf: [u8; 32],
+        new_leaf: [u8; 32],
+        i: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let proof = full_proof::<H, MAX_HEIGHT>(proof);
+        // let mut updated_proof = proof.clone();
+
+        for j in 0..(self.current_index as usize) + 1 {
+            println!(
+                "iteration j: {j}, root: {root:?}, roots[j]: {:?}",
+                self.roots[j]
+            );
+            if root == self.roots[j] {
+                self.replace_leaf_inner(old_leaf, new_leaf, i, j, &proof)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Appends a new node to the tree.
+    pub fn append(&mut self, mut leaf: [u8; 32]) -> Result<(), HasherError> {
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+        let mut intersection_node = self.rightmost_leaf;
+        let intersection_index = self.rightmost_index.trailing_zeros() as usize;
+
+        if self.rightmost_index == 0 {
+            println!("FIRST INSERTION");
+            // NOTE(vadorovsky): This is not mentioned in the whitepaper, but
+            // appending to an empty Merkle tree is a special case, where
+            // `computer_parent_node` can't be called, because the usual
+            // `self.rightmost_index - 1` used as a sibling index would be a
+            //  negative value.
+            //
+            // [spl-concurrent-merkle-tree]()
+            // seems to handle this case by:
+            //
+            // * Valitating a proof.
+            // * Performing procedures which usually are done by `replace_leaf`
+            //   algorithm.
+            //
+            // Here, we just call compute the new root and call `replace_leaf`
+            // directly.
+            leaf = H::hashv(&[&leaf, &H::zero_bytes()[1]])?;
+            let proof = self.rightmost_proof.clone();
+            self.replace_leaf_inner(H::zero_bytes()[0], leaf, 0, 0, &proof)?;
```

Comment:
we could also just start with index 1 instead and leave index zero the zero values

Reply:
I'm not sure I'm following your idea.

`H::zero_bytes()[0]` is passed here as `old_leaf` parameter - this way we make it explicit that we replace the first leaf, being zero at that time, with a new one. Feels intuitive (at least for me :D) and I wouldn't like to obfuscate it.

But I also feel like I'm misunderstanding you. Which parameter would you like to replace in this call?

Reply:
Ah, OK, I probably get it now. You mean if we could just leave the 1st leaf/subtree (index zero) always as zero byte? That would be inconsistent with spl-concurrent-merkle-tree, so we couldn't go forward with your idea of testing both implementations.

(But maybe I'm still misunderstanding you :P )

Reply:
I see makes sense.